Conda packages and image test-driven text removal for Miniconda on Appveyor for #744 and #690

nickpowersys opened the pull request:
### Summary

To address the **failing imports** in Windows VMs with Miniconda:
- created a **conda environment with conda packages** in .appveyor.yml, which fixes #744.

The  **Matplotlib backend configuration** has been addressed by #862. mpl.use('agg') has been working as intended, as explained in that PR.

To minimize the probability of image comparison tests failing in general:
- removed x and y labels
- changed default value of remove_legend to True
- updated baseline images

Before removing the labels and legends, approximately 25% of Windows image comparison tests have required an elevated value of tol in order to pass (227 tests have the check `if sys.platform == 'win32'`), out of 894 image comparison tests.

When `'win32'` is changed to `'wwwww'` purely for analysis purposes, ignoring the Windows-specific `tol` values, the number of failing tests with regular Python on Appveyor is at 24, and with Miniconda, the number is 31. Therefore, **the number of Windows tests with RMSE values >0.1 has dropped roughly 90%**. Follow-up efforts can remove the instances of Windows-specific `tol` values.

Builds for above analysis:
PyPI/pip Python: https://ci.appveyor.com/project/nickpowersys/yellowbrick/builds/25047957
Anaconda: https://ci.appveyor.com/project/nickpowersys/yellowbrick/builds/25049306

**To create the conda environment**, I used tests/requirements.txt.
The conda packages were determined by the conda solver.
The name of the environment is yellowbrick. It is activated directly following the installation of the packages so that any further commands with the environment are done from the new environment.

In general, it can be expected that conda environments are used primarily with conda packages, and therefore using conda packages in Miniconda builds can help to verify conda users'  and developers' experience through tests using these environments, including those involving the backend.

bbengfort commented:
@nickpowersys thank you so much for taking a look at this and diving in with the update! We don't have very many Windows/Anacondas users on the core contributor side, so we struggle with supporting it and really appreciate your PR.  Also, I apologize if I end up asking basic questions for clarification as a result. 

To summarize the solution to make sure we're on the same page - if using miniconda (identified by the `MINICONDA_VERSION` envvar), then running the tests in a conda environment ensures that the DLLs associated with the miniconda install are correctly loaded?  If not using miniconda, then simply proceed with the pip/python mechanism, which has not been changed. The `agg` backend cropped up when making this change as per #690 - so the solution was to modify the configuration as per that discussion. 

It looks like the tests are now failing due to increased RMSE in the conda environment. The failing tests all appear to have an elevated tolerance if `sys.platform == win32`. When we previously had access to a Windows machine we took a look at the image comparisons and found that there were font and aliasing issues that may have contributed to the error; would you be willing to do a minor spot check on the conda environment and see if that's also the case? Do you know if it is possible to check from within Python whether or not the Python version you're using is anaconda/miniconda? 

To answer your questions:

- We should probably handle the failing tests in this PR as they seem to be related to the conda environment; I'm hoping there is an easy configuration solution.
- Documentation about locating the configuration would be helpful; I'm also willing to add the `MPLBACKEND=agg` to the tests to supersede all configurations, as suggested in #690. 

Thank you again for taking a look into this - it's a relief that we're moving forward and may be able to test miniconda on AppVeyor again!

nickpowersys commented:
@bbengfort Thank you for your feedback! I'm glad this is useful. About the conda environment, imports and DLLs, and using pip for non-conda VMs, yes.

I would be happy to look at aliasing that may be occurring. It may take me until next weekend to investigate.

It is possible to determine within Python whether a regular Python or Anaconda Python is being used. I could reply with more detail after ensuring the method is reliable.

nickpowersys commented:
I tried uncommenting several settings for antialiasing in matplotlibrc, which did not result in any changes in the number of tests failing (16) on a Windows machine with a conda environment. I looked at a pair of images (test_binary_probability.png) and could not see aliasing.

Is there a configuration change you would recommend? It is possible to detect whether the active Python is in a conda environment using Python, based on whether a directory specific to conda installations ('conda-meta') is part of the active environment path.

Using an environment variable for the backend seems straightforward to apply for testing purposes.

nickpowersys commented:
@bbengfort In addition to the update on the image comparison, I'm merging the changes from develop.

nickpowersys commented:
@bbengfort Would you recommend an approach of xfails for the failing tests in conda environments for now? For this PR, I would resolve those that are failing on Windows (there are 16). The RMSE are well below the tol of 10 mentioned in #393. I would address backend tests with MPLBACKEND=agg.

To understand possible causes for the image comparison failures, I ran builds on Travis for Python 3.6 and 3.7 Miniconda environments with conda packages on Ubuntu Xenial, and image comparison tests failed at a higher rate than on Windows (those were the only types of failing tests on Linux).

Therefore, there seems to be something associated with conda packages or dependencies that is causing the comparisons to fail. The xfails would be based on the Anaconda interpreter being active.

nickpowersys commented:
There is a possibility that Freetype is the reason. I can investigate this before going to xfails.

https://github.com/DistrictDataLabs/yellowbrick/issues/690#issuecomment-487999218

nickpowersys commented:
While there are some differences in fonts between baseline and test images in failing tests, text is not found in the baseline and test images in general. It seems there should not be text in the test images, based on existing code that removes it prior to comparison tests, and as seen in baseline and actual images under passing tests. I will need to investigate further to see why there is text in failing tests.

Before comparisons are done, the ImageComparison class has a cleanup() function that removes the title, legend and text on the x, y and z axes with a NullFormatter [https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/base.py](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/base.py)

As an example passing test, the baseline images in [https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/baseline_images/test_classifier/test_class_prediction_error/test_class_prediction_error_quickmethod.png](https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/baseline_images/test_classifier/test_class_prediction_error/test_class_prediction_error_quickmethod.png) have all text removed. The images in the corresponding actual_images directory also do not have text.

bbengfort commented:
@nickpowersys I'm sorry it's taken me so long to respond; we've been bogged down in PRs and issues and many of us were at PyCon so your messages got lost in all of the GitHub notices; I apologize. 

I really do appreciate you taking a look at this for us. We really do want to support our Windows/Anaconda users but unfortunately, none of us are Windows/Anacondas users, so it makes it tricky. 

To answer some of your earlier questions:

- We do modify the `rcParams` in `yellowbrick.style` - perhaps these rcParam modifications are overriding what you're doing with anti-aliasing? I'd be happy to take a look if you mention the specific parameters. 

- Even though we do try to strip out as much text as possible, occasionally there is still some text that gets in. E.g. the numeric labels in classification reports. My sense is that text is probably where things are going wrong.  

- Thank you for investigating freetype - I do recommend the use of `xfail` if a conda environment is detected. Of course, we are hoping for a better solution but having many tests with conda is better than having None! 

How can I assist you in getting this to the end with the xfails? Do you want to investigate the style overrides for anti-aliasing? Also, sorry if I missed this, but do you have the tests passing locally on your windows machine? I'd be interested in the difference between the baseline and actual images on Windows if you were looking at them.

bbengfort commented:
@nickpowersys - on #690 there was some discussion about using the testing version of matplotlib to unify the freetype dependency with anaconda. It can be installed using conda-forge as follows:

```
conda install -c conda-forge/label/testing matplotlib
```

The version may be a bit behind, (currently 3.0.2) but it might be worth it so we don't have to xfail all these tests. Would you be willing to try this in this PR?

nickpowersys commented:
During the weekend, I found that the reason for text appearing in the images when it should not have was because although tick marks and the title were being removed, the x and y labels were not (NullFormatter addresses ticks but not labels). In order to reduce the failing tests as much as possible (the failing Windows Miniconda tests were at 16 after addressing the backend for conda), I will update the rest of the baseline images on my current branch to remove the text using Linux.. I can see how many failing tests remain. 

Legends are not being removed by default. It would just take a change of that default value to True to remove them across all image comparisons, but I am leaving that in place for now since I assume there was a reason it is False. It may be useful to set it to True for those failing tests where the legend contains substantial text. If you have any thoughts on removing legends either way, please let me know. I could look at Freetype if tests were still failing and there was a need to preserve legend text.

The change created a situation where a significant number of existing baseline images that exist have label text, while the actual images do not. This caused failing tests on Travis. I have set up an Ubuntu laptop so that I can generate them with Linux using a regular Python installation, and have them be the baseline images. Then, I can see if there are any remaining aliasing issues.

Aside: before using Windows on some projects, I used Ubuntu until hardware issues affected my previous laptop. Otherwise, I do development on a Mac.

bbengfort commented:
@nickpowersys thanks for sticking with this! I realize that any changes to the image similarity tests generally affects every single baseline image. If you need help regenerating the images, please let @ndanielsen or I know!

nickpowersys commented:
@bbengfort I 

- removed x and y labels
- removed legends where the tests were still failing
- updated the baseline images

Then, there were 5 failing tests in tests/test_features/test_jointplot.py on the Miniconda builds. Within that test file, I specified higher tol values (between roughly 2 and 5) if the interpreter was in a conda environment.

nickpowersys commented:
The diffs created through test_jointplot.py show some vertical gridlines that were not aligned, and some subtle aliasing otherwise. This also fixes #690.

nickpowersys commented:
In the process of finalizing the .appveyor.yml Miniconda-related edits, I set the MPLBACKEND to 'agg', using SET and SETX. The environment variable did take effect, but for some reason, was not actually used during the tests. Therefore, the backend is set in the matplotlibrc file.

nickpowersys commented:
@ndanielsen Do you know what you were assigned to address on the PR? At this point, if there is an intent to address the conda environments on Linux, not just Windows, I have run tests for conda environments on Ubuntu.

I saw that there are differences in the text between it and the regular Python text that lead to roughly 70 image comparison failures.

There are some cases in which VisualizerGrids have subplots, and only the last subplot has had its text removed. In other cases without subplots, there are legends with differences in text.

The legends are easy to remove and resolve and I could generate new baseline images based on these tests. I can mark the other failing tests as xfail where there is an Ubuntu platform with a conda environment.

ndanielsen commented:
@nickpowersys thanks for pinging me. I'm getting caught up on this issue at the moment. I plan to deep dive into this over the next couple days.

nickpowersys commented:
@ndanielsen @bbengfort @rebeccabilbro I wanted to let you know that there is a working build for Appveyor including regular and Miniconda Python VMs. The additional commits beyond the previous one from Friday are minor, involving only two tests that were failing with Miniconda. The build success may not appear in this PR at the moment due to a Slack authentication token issue, but it is here:

[https://ci.appveyor.com/project/nickpowersys/yellowbrick/builds/24650172](url)

The develop branch has moved ahead, and there are merge conflicts that I'm not able to view.

@rebeccabilbro Do you want to keep this PR's scope to addressing Appveyor/Windows, then move beyond it with a separate PR to address the tests that are not addressed for non-Windows conda environments? I understand that there is a goal of supporting Miniconda builds on Travis.

I created a travisconda branch independently of appveyorminiconda to see the Travis status. Based on running test builds, there are at least 20-25 test files with 73 image comparison tests that can be resolved by removing legends (which are not causing failing tests on Windows), or, where there are additional  elements in the plots that are not easily removed, or other differences, that can be marked xfails on conda Linux/Mac environments.

nickpowersys commented:
Running with regular Python 3.6 and 3.7, the Travis builds for Xenial are successful:

https://travis-ci.org/nickpowersys/yellowbrick/builds/534519030

nickpowersys commented:
**Freetype**: I ran a Travis build after installing the testing version of matplotlib 3.0.2 into a conda environment. However, I saw no change in the number of tests passing/failing.

**Merging**: One cause of the merge conflicts is the updated .png files.

In order to have the builds show up in the PR with merged commits from develop, LGTM indicates I need to address the appveyorminiconda branch being in a detached head state.

Since I have not dealt with the detached head situation and have not found discussion of how best to handle it for YB in the repo, I would very much appreciate some guidance on how to proceed! If it would speed things up, I could chat in Slack. I would also like to ensure that Slack authentications are working, if you can provide any relevant steps.

Thank you!

nickpowersys commented:
**Freetype**: I ran a Travis build after installing the testing version of matplotlib 3.0.2 into a conda environment. However, I saw no change in the number of tests passing/failing.

**Merging**: One cause of the merge conflicts is the updated .png files.

In order to have the builds show up in the PR with merged commits from develop, LGTM indicates I need to address the appveyorminiconda branch being in a detached head state.

Since I have not dealt with the detached head situation in the context of a PR and have not found discussion of how best to handle it for YB in the repo, I would very much appreciate some guidance on how to proceed! If it would speed things up, I could chat in Slack. I would also like to ensure that Slack authentications are working, if you can provide any relevant steps.

Thank you!

ndanielsen commented:
@nickpowersys Thanks for your patience and comments over the weekend. Please accept my apologies that we were unable to immediately respond to them.  

As I mentioned in https://github.com/DistrictDataLabs/yellowbrick/pull/823#issuecomment-493715701, I'm doing a deep dive on these windows testing issues today, and hopefully I can offer some guidance soon.

Definitely want to acknowledge that you've done a ton of work on this issue, and we're grateful for it.

ndanielsen commented:
@nickpowersys I spent a good chunk of today diving into several of the issues that this PR addresses. Also another similar issue popped up in this PR: https://github.com/DistrictDataLabs/yellowbrick/pull/854 due to changes in matplotlib==3.1.0 where the backend seems like it can no longer be set programmatically as is done here: https://github.com/DistrictDataLabs/yellowbrick/blob/e68e28ece6032898878c318900661bbf1303baae/tests/__init__.py#L24 

The `matplotlibrc` line in your appveyor config pointed me to solution for the issue in #854 

I think that the best way to manage the backend for testing is creating a  `matplotlibrc` file in the project root and setting `backend : agg` inside as [described here in the docs ](https://matplotlib.org/users/customizing.html#the-matplotlibrc-file). This might also be a more transparent way to set other attributes for testing, but more research on that would be required.

When I did this on one of my branches, tests were passing on travis, appveyor, my macbook, my ubuntu box, and windows10 in a virtualbox. (the stars seemed to be aligned on this)  With matplotlib==3.1.0, there were some failing image comparison tests in `test_class_balance` and `test_rfecv` but that is expected behavior when dependencies get a bump up. When I generated new images, there were no problems. 

What do you think of this as a solution? I know that you've done a ton of work on these issues, I want you to get your credit as a contributor. If you're okay with this do you mind opening a separate PR for #690 adding a `matplotlibrc` file in the project root and also allowing `matplotlib==3.1.0` plus generating any new baseline images? 

After that, let's put our heads together on highly focused PR on #744 and address the miniconda issues. I like your clever solution for removing the x, y labels. I've also been tinkering around with a separate `windows_tolerance` metric for when the test suite is run on windows machines. I think that it is a better solution than an expected fail. Also when I ran the test suite locally on my ubuntu box in a miniconda, the image failure diff was the result of the images being rendered a few pixels off. 

Thanks again for your efforts on these issues. A robust testsuite and cross-platform CI is highly important for a growing library like yellowbrick.

nickpowersys commented:
Thank you very much for your analysis. I hope to reply this evening. I am
tied up in a conference until the early evening.

Nick

On Tue, May 21, 2019, 12:04 AM Nathan <notifications@github.com> wrote:

> @nickpowersys <https://github.com/nickpowersys> I spent a good chunk of
> today diving into several of the issues that this PR addresses. Also
> another similar issue popped up in this PR: #854
> <https://github.com/DistrictDataLabs/yellowbrick/pull/854> due to changes
> in matplotlib==3.1.0 where the backend seems like it can no longer be set
> programmatically as is done here:
> https://github.com/DistrictDataLabs/yellowbrick/blob/e68e28ece6032898878c318900661bbf1303baae/tests/__init__.py#L24
>
> The matplotlibrc line in your appveyor config pointed me to solution for
> the issue in #854
> <https://github.com/DistrictDataLabs/yellowbrick/pull/854>
>
> I think that the best way to manage the backend for testing is creating a
> matplotlibrc file in the project root and setting backend : agg inside as described
> here in the docs
> <https://matplotlib.org/users/customizing.html#the-matplotlibrc-file>.
> This might also be a more transparent way to set other attributes for
> testing, but more research on that would be required.
>
> When I did this on one of my branches, tests were passing on travis,
> appveyor, my macbook, my ubuntu box, and windows10 in a virtualbox. (the
> stars seemed to be aligned on this) With matplotlib==3.1.0, there were some
> failing image comparison tests in test_class_balance and test_rfecv but
> that is expected behavior when dependencies get a bump up. When I generated
> new images, there were no problems.
>
> What do you think of this as a solution? I know that you've done a ton of
> work on these issues, I want you to get your credit as a contributor. If
> you're okay with this do you mind opening a separate PR for #690
> <https://github.com/DistrictDataLabs/yellowbrick/issues/690> adding a
> matplotlibrc file in the project root and also allowing matplotlib==3.1.0
> plus generating any new baseline images?
>
> After that, let's put our heads together on highly focused PR on #744
> <https://github.com/DistrictDataLabs/yellowbrick/issues/744> and address
> the miniconda issues. I like your clever solution for removing the x, y
> labels. I've also been tinkering around with a separate windows_tolerance
> metric for when the test suite is run on windows machines. I think that it
> is a better solution than an expected fail. Also when I ran the test suite
> locally on my ubuntu box in a miniconda, the image failure diff was the
> result of the images being rendered a few pixels off.
>
> Thanks again for your efforts on these issues. A robust testsuite and
> cross-platform CI is highly important for a growing library like
> yellowbrick.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DistrictDataLabs/yellowbrick/pull/823?email_source=notifications&email_token=ABCSFOUYIAASSOGXGXOHS5DPWN7E3A5CNFSM4HHMN2O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV2YHKA#issuecomment-494240680>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCSFOSWWZBEGHJULHMZA23PWN7E3ANCNFSM4HHMN2OQ>
> .
>

bbengfort commented:
@ndanielsen and @nickpowersys I know you guys are cranking away at this and I want to say thank you to you both! @ndanielsen is the original author of our image similarity tests ported from matplotlib, so you're in good hands now @nickpowersys - good luck and let me know if you need me to chime in at any point!

nickpowersys commented:
@ndanielsen I was in a conference that still has a half day remaining, but
I can respond to your deep dive.

I can open a PR that sets the backend through the matplotlibrc file. For
testing through builds, I could put the file into the working directory,
the first of the possible locations listed in the docs, essentially
creating it as part of the build script.

As part of that PR, I would also address the backend for local user
testing, including the issue in 854. I think it may be possible to do this
through os.environ['MPLBACKEND']='agg' as indicated in 690. I can verify
that it works on my local machines. This could be done by the user through
this method or the matplotlibrc, if they are prompted by a short message in
the assertion test for the backend in the event that it fails, indicating
that this approach or matplotlibrc may be used as indicated in the docs.

I hope it would be okay if we keep the scope and approach for 823 to
specifying the appveyor.yml in order to have a conda environment and
packages installed, and then use xfails with the failing tests. The number
of xfails was limited to 5 after the x and y labels were removed. It would
still require me to submit a commit or two to ensure that all tests and
baseline images that are now part of develop are handled.

I agree that there is a need to have another approach to the differences
that occur with image comparison tests beyond xfails, like the one you
tested, then undo the xfails. I would be glad to huddle with you on it.
Does this sound good?


On Tue, May 21, 2019 at 1:58 PM Benjamin Bengfort <notifications@github.com>
wrote:

> @ndanielsen <https://github.com/ndanielsen> and @nickpowersys
> <https://github.com/nickpowersys> I know you guys are cranking away at
> this and I want to say thank you to you both! @ndanielsen
> <https://github.com/ndanielsen> is the original author of our image
> similarity tests ported from matplotlib, so you're in good hands now
> @nickpowersys <https://github.com/nickpowersys> - good luck and let me
> know if you need me to chime in at any point!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DistrictDataLabs/yellowbrick/pull/823?email_source=notifications&email_token=ABCSFOQ7WE6VAFN75PE3VQLPWRA5ZA5CNFSM4HHMN2O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV43MWA#issuecomment-494515800>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCSFOWVVHYZ6WZDPW3TSGLPWRA5ZANCNFSM4HHMN2OQ>
> .
>

ndanielsen commented:
@nickpowersys thanks for the response. I hope that you are enjoying the conference. They definitely take a lot out of you. I feel like I'm just starting to get my normal levels of energy back after spending about a week at pycon. =)

I am block time to work on this and other yellowbrick testing related issues on Friday.

For setting the MPL backend and related configuration for testing, I think that part of the solution needs to keep the `contributor experience` in mind in addition to CI configuration. Whatever the end solution is, I think that we should strive to make running the testsuite on mac, linux and windows as easy as possible along the lines of this workflow:

```
git clone http://github.com/../yellowbrick.git
cd yellowbrick
python3 -m venv yb-dev && source yb-dev/bin/activate 
pip install -r tests/requirements.txt
make test
```
I think one of the first goals is to have a painless `contributor experience` and we can extend the CI solutions(s) around that, but I'm definitely not tied to any particular solution. =)

As for structuring this effort, let's try to keep the PRs small, focused and minimally complex so that we can incrementally add value. The best approach might be to slowly "paint these problems into a corner", and (re)-build up cross-platform testingsuite support is the way to go.

Again, thanks for your continued follow up on these issues. This is probably one of the most important issues on the board as it directly impacts how easy it is to maintain the project and make contributions to the project easy. 

Enjoy the rest of your conference. 🥇

ndanielsen commented:
@nickpowersys I agree with your suggestion for limiting this just to `appveyor.yml`.  

```
I hope it would be okay if we keep the scope and approach for 823 to specifying the appveyor.yml in order to have a conda environment and packages installed, and then use xfails with the failing tests. The number of xfails was limited to 5 after the x and y labels were removed. It would still require me to submit a commit or two to ensure that all tests and baseline images that are now part of develop are handled.
```

nickpowersys commented:
@ndanielsen I pulled develop through and including the PR for the added rankd test (866) then did an interactive rebase (I confirmed that PR 866 in the git log for the appveyorminiconda branch). I tested on Linux pyenv locally and Windows conda locally, and would like to test on the builds including Miniconda for Appveyor.

The updates reflect the removal of both x and y labels, as well as legends. After updating the baseline images on Linux pyenv, I was able to remove small number of xfails (~5) where they had been used for conda environments on Windows.

Before proceeding further, I wanted to confirm that it will be okay to push with a regular force to origin/appveyorminiconda (not force-with-lease). I understand that it should be okay because the current origin/appveyorminiconda has not been pushed to develop. I would do `git push origin appveyorminiconda -f`. Whatever you recommend, I will do.

The only baseline_image that I did not understand and did not update (when I tried, the image comparison did not fail as intended) is test_scatter_image_fail. The test is passing with the existing baseline image, which still has labels and a legend.

Thank you!

nickpowersys commented:
Out of the original 15 commits, I have three commits that cover

- conda environments in .appveyor.yml
- text removal from baseline images
- an escape sequence warning

ndanielsen commented:
@nickpowersys Thanks for your follow up on this.

There are two main paths that we could take here:

- A force push to your local branch after rebasing shouldn't mess anything up in the `yellowbrick:develop` branch
- You can start a new PR and cherry-pick (or manually add in) those commits into that PR just reference this PR for history keeping.

I lean slightly towards the second one just because I have a tendency to like clean code and compact PRs, but the first option is viable also.

As for the `test_scatter_image_fail` comparison test, I wrote that as a true positive case of when the test was failing. As the testsuite has now matured, we can remove that.

Outside of that, thanks for your great efforts and follow through on this!

nickpowersys commented:
This is the error in in Appveyor:
> _ ERROR collecting tests/test_base.py _____________________
<frozen importlib._bootstrap>:888: in _find_spec
    ???
E   AttributeError: 'AssertionRewritingHook' object has no attribute 'find_spec'
During handling of the above exception, another exception occurred:
C:\Python36-x64\lib\site-packages\py\_path\local.py:701: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:951: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:890: in _find_spec
    ???

I think it might have something to do with `from yellowbrick.base import *` not being used. In my IDE, the line is grayed out. I can try commenting it out with a Windows pyenv environment.

> import unittest

from yellowbrick.base import *
from yellowbrick.features.base import *
from sklearn.base import BaseEstimator, TransformerMixin

nickpowersys commented:
I am also running build on a branch created from appveyorminiconda in its current state, and specifying `matplotlib==3.0.0`.

If Appveyor builds finish, then I will remove `matplotlib ... !=3.0.0` from tests/requirements.txt. and `requirements.txt`
https://ci.appveyor.com/project/nickpowersys/yellowbrick

Travis build completed:
https://travis-ci.org/nickpowersys/yellowbrick/builds/540523850

nickpowersys commented:
There are no other changes for the PR relating to Appveyor and Miniconda as of now.

ndanielsen commented:
For the record, here's the actual image drawn in ubuntu and the error diff:

![test_dispersion_plot_annotate_docs](https://user-images.githubusercontent.com/2944777/58842346-977be500-8622-11e9-8f91-5fdf660463db.png)

![test_dispersion_plot_annotate_docs-failed-diff](https://user-images.githubusercontent.com/2944777/58842354-9fd42000-8622-11e9-9fd0-8f3e0ce420be.png)

nickpowersys commented on tests/base.py:
```diff
@@ -171,8 +171,9 @@ class ImageComparison(object):
 
-    def __init__(self, stack, visualizer=None, ax=None, tol=0.01, windows_tol=0.01, ext=".png",
-                 remove_ticks=True, remove_title=True, remove_legend=False):
+    def __init__(self, stack, visualizer=None, ax=None, tol=0.01, 
```
I can bump up the tol to 25.5, because it is not a significant change. It seems like it may be due to a difference in colors, because the diff is so extensive. I wonder if we may be using different versions of a package, or if there is some difference in configuration. I ran the tests again locally for Ubuntu with regular Python and conda, and removed the tol=25 to see if it would fail, but it passed in both cases. I'm using matplotlib 3.1.0, and Python 3.6.8.

Sorry for not thinking about the update to the docs, will do!

ndanielsen replied:
For that flaky test in ubuntu, I agree to bump it up to 25.5. I'm on ubuntu 18.0 ran it on python 3.6.8 and 3.7.3 failing in both. It's hard to say what is causing it without doing a deep dive.

After updating the docs, this should be good to merge in! Thanks again for the great work on this.

nickpowersys commented on tests/test_features/test_jointplot.py:
```diff
@@ -47,6 +48,13 @@
+            os.path.exists(os.path.join(sys.prefix, 'conda-meta')))
+
+
+win_tol = 5.5 if is_win_conda_env() else None
```
Yes, when I ran on conda on Linux most recently, there were 36 image comparisons failing. The extent of the differences previous to this week was something I discussed in https://github.com/DistrictDataLabs/yellowbrick/pull/823/#issuecomment-493755156

I would like to get the PR merged, as I'm sure you would! However, just before finalizing it, something occurred to me that I wanted to bring up. This might be something you already considered, or may want to discuss with others. It just became more clear in having to handle the multiple Windows builds, but it is more general than Appveyor/Windows comparisons.

Since part of the goal is to support conda across Windows and Linux (and I'm guessing Mac on Travis generally), I'm wondering if having a single `osenv` variable  might be the easiest way to handle the default tols for different OS/Python environment types combinations, as well as overrides. This is of course subjective, so it may be that for maintenance purposes and the contributor experience, a more coarse-grained OS-only set of tols are more appropriate.

Any approach has to allow maintainers to easily tune the default tols. There is the option of either just using an OS-based value for tol (`ostol`), or a more granular OS- and Python version-based tol (`osenvtol`).

If overrides were needed at the module or test level, an `osenv` variable could be imported for example (consistent with your suggestion to move the is_conda_env() out of the single module to base.py). This would allow for the maximum granularity without applying the same values across both regular Python and conda on the same OS.

**tol keyword arg(s)**: This is just my thought, but in either case, having just the single tol kwarg for overrides seems like it may require less effort to code in, and still work across OS platforms as well as Python environments.

Once there is an OS-specific tol, it becomes more effort to think "what is just plain tol for now? Is it best to handle conda separately, or just apply an OS-level tol, which may be higher than is needed for at least one of the Pythons (PyPI/conda)?" My current approach of using both the tol and windows_tol to handle both regular Python and conda having their own tols is definitely akward. The only good thing is that it applies more granularity to tols, based on the respective environment-specific builds.

If there is just one way to approach it by convention, whatever the convention, that can be applied to the PR.

The code that would allow for specifying default OS+py type tols would look something like this, if we wanted to allow for up to six default tols:

Within tests/base.py:

```

def is_conda_env():
	return os.path.exists(os.path.join(sys.prefix, 'conda-meta'))

def osenv_str():
	platform = {'win32': 'win', 'linux': 'lin', 'darwin'[:5]: 'mac'}
	py_or_conda = 'py' if not is_conda_env() else 'conda'
	return platform[sys.platform[:5]] + py_or_conda

# Configuration by maintainers

tol = 0.1

env_tols = {'linpy': tol, 'linconda': 0.1,
            'winpy': 0.1, 'winconda': 0.1,
             'macpy': 0.1, 'macconda': 0.1}

# Automatic based on configuration
os_tols = {'linux': max([v for k, v in env_tols.items() if 'lin' in k]),
            'win32': max([v for k, v in env_tols.items() if 'win' in k]),
            'darwin'[:5]: max([v for k, v in env_tols.items() if 'mac' in k])}


osenv = osenv_str()
osenvtol = env_tols[osenv]
ostol = os_tols[sys.platform[:5]]
```

For ImageComparisonclass, the `tol` could be specified by the maintainers like this:

`def __init__(tol=osenvtol)`

or like this:

`def __init__(tol=ostol)`

If there is no need for different OS/Python type tols in a module, an image comparison is the default
`self.assert_images_similar(visualizer)`

Or, if a test fails in a module despite the default tol from `__init__`, this is possible:

```
from tests.base import osenv, tol  # tol is 0.1, but no need to remember it

if osenv == 'winpy': tol = 0.5   # OR tol = 0.5 if osenv == 'winpy' else 0.1
self.assert_images_similar(visualizer, tol=tol)
```

Or, if there are 2 or more builds failing for a given module or test, something like this is possible at the top of a module or at the test level:

```
from tests.base import osenv, tol

# For example only
env_tols = {'winpy': 0.5, 'winconda': 5.5, 'linconda': 6.0}
if osenv in env_tols: tol = env_tols[osenv]
self.assert_images_similar(visualizer, tol=tol)
```

ndanielsen replied:
@nickpowersys thanks for the thoughtful and detailed analysis on tolerances across operating systems. 

My sense is that this is a bigger conversation. In theory, having different tolerances across osenvs is suboptimal as we hope that the library could perform identically in each osenv. We probably should do a deeper dive into why the images produced are much different and draw inspiration from the approach that matplotlib takes to this very issue.

I recommend that you open up an [issue](https://github.com/DistrictDataLabs/yellowbrick/issues/new/choose) essentially with your proposal here. Perhaps you could list out the state of the image similarity tests across each of the osenvs to help steer the conversation and eventual work. 

For the moment thought, let's close out this PR knowing that we have windows conda and venv coverage in a much stronger place. It'll allow many other contributors to directly benefit from this work while the next task at hand is getting to the bottom of these bigger issues. 

What do you think about this approach?

nickpowersys commented:
If I could merge from develop, then I could regenerate the baseline images from test_draw.py.

ndanielsen commented:
@nickpowersys I'll see what I can do on this tomorrow.

ndanielsen commented:
@nickpowersys I think that I figured this one out and got it passing in a [PR on my personal fork](https://github.com/ndanielsen/yellowbrick/pull/6).

Following the github instructions... 

```
 $ git checkout -b nickpowersys-appveyorminiconda develop
 $ git pull git://github.com/nickpowersys/yellowbrick.git appveyorminiconda
 $ git checkout develop
 $ git merge --no-ff nickpowersys-appveyorminiconda
```

the only change that I was that instead of `git push origin develop`, I merged it back into the feature branch like this:

```
 $ git checkout appveyorminiconda
 $ git merge develop
```

Followed by a `git push` as appropriate.

Hope this helps!

ndanielsen commented on tests/base.py:
```diff
@@ -279,7 +293,11 @@ def cleanup(self):
-        if self.remove_legend:
+        if self.remove_labels:
+            self.ax.set_xlabel("")
+            self.ax.set_ylabel("")
```
Just note that this solves the xlabel, ylabel issue that has popped up in mtpl 3.x

nickpowersys commented:
@ndanielsen Thanks for providing this git-fu! I just want to make sure I
understand the current context of the conflict.

When I did my last push to appveyorminiconda, there was a conflict with a
.png file test_manual_legend.png that I thought was expected, given that my
push had the same baseline image as a very recent PR committed to develop
based on test_draw.py, just with the text removed. The Github UI said and
says now that the way to resolve it was for someone with write access to
accept my version of the .png file (or not).

I can make sure there is no conflict or issue in tests/base.py as well.

After that, I could pull the updated version of appveyorminiconda and
generate new baseline images for test_draw.py, which would also generate
conflicts (they would be new to appveyorminconda).

If you are saying that I can avoid the small set of .png conflicts by
following these steps starting from my current local git state (OR that you
plan to resolve them all in a single batch, instead of resolving this
single .png conflict now, and the rest later), then I will go ahead later
today..thanks for clarifying!

On Thu, Jun 6, 2019, 10:33 AM Nathan <notifications@github.com> wrote:

> *@ndanielsen* commented on this pull request.
> ------------------------------
>
> In tests/base.py
> <https://github.com/DistrictDataLabs/yellowbrick/pull/823#discussion_r291244069>
> :
>
> > @@ -279,7 +293,11 @@ def cleanup(self):
>                  except AttributeError:
>                      continue
>
> -        if self.remove_legend:
> +        if self.remove_labels:
> +            self.ax.set_xlabel("")
> +            self.ax.set_ylabel("")
>
> Just note that this solves the xlabel, ylabel issue that has popped up in
> mtpl 3.x
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DistrictDataLabs/yellowbrick/pull/823?email_source=notifications&email_token=ABCSFORKMHH6H2Q7ONLF6ULPZEU4NA5CNFSM4HHMN2O2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2ZZHMA#pullrequestreview-246649776>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCSFOQF55EYLDHVJ7ZMM6DPZEU4NANCNFSM4HHMN2OQ>
> .
>

ndanielsen commented:
@nickpowersys I've encountered this #git-problem a couple times where there are protected branches. 

I think that the problem context is simple that [tests/baseline_images/test_draw/test_manual_legend.png](
https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/baseline_images/test_draw/test_manual_legend.png) was updated ~2 days ago, and your current branch doesn't have those changes. 

The solution is to merge in the `development` branch with a special git-fu and get tests passing without merge conflicts. As this PR has 197 files changed (large scope) and there are other PRs under active development, there is a lot of overhead and potential conflicts in keeping this open. 

Outside of that, thanks again for your work on this PR and others. I'm starting to actively reference some of the changes in other PRs with other contributors and it's solved specially matplotlib 3.x problems that we didn't know that we had.

nickpowersys commented:
I realize now where I missed merging in the upstream develop commits into
my remote appveyorminconda before pulling from the remote, and it's clear
now how to avoid this situation going forward. I will follow your steps to
resolve this.

I am moving in-town for the next hour or so, but after that, my immediate
priority is to close it out. Thank you!

On Thu, Jun 6, 2019, 12:49 PM Nathan <notifications@github.com> wrote:

> @nickpowersys <https://github.com/nickpowersys> I've encountered this
> #git-problem a couple times where there are protected branches.
>
> I think that the problem context is simple that
> tests/baseline_images/test_draw/test_manual_legend.png
> <https://github.com/DistrictDataLabs/yellowbrick/blob/develop/tests/baseline_images/test_draw/test_manual_legend.png>
> was updated ~2 days ago, and your current branch doesn't have those changes.
>
> The solution is to merge in the development branch with a special git-fu
> and get tests passing without merge conflicts. As this PR has 197 files
> changed (large scope) and there are other PRs under active development,
> there is a lot of overhead and potential conflicts in keeping this open.
>
> Outside of that, thanks again for your work on this PR and others. I'm
> starting to actively reference some of the changes in other PRs with other
> contributors and it's solved specially matplotlib 3.x problems that we
> didn't know that we had.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DistrictDataLabs/yellowbrick/pull/823?email_source=notifications&email_token=ABCSFOSBMYEVJXLTSSA3RB3PZFE4HA5CNFSM4HHMN2O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXDUNOA#issuecomment-499599032>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABCSFOUQERDCYVHANFSR5LDPZFE4HANCNFSM4HHMN2OQ>
> .
>

nickpowersys commented:
I went through the steps 1-6 (through `git merge develop`), then ran `make test` locally on Linux. 759 tests passed, and none failed. Then `git push`.

nickpowersys commented:
:loudspeaker: A huge thank you @ndanielsen!

ndanielsen commented:
@nickpowersys thanks for your big efforts on this! We sincerely hope to keep you as a contributor on yellowbrick!
